domain/schema/model/sql/0020-network.sql

```diff
@@ -14,9 +14,304 @@ CREATE UNIQUE INDEX idx_cloud_service_application
+    FOREIGN KEY (config_type_id)
+    REFERENCES ip_address_config_type (id),
+    CONSTRAINT fk_ip_address_scope
+    FOREIGN KEY (scope_id)
+    REFERENCES network_address_scope (id)
+);
+
+CREATE TABLE ip_address_dns_server_address (
+    address_uuid TEXT NOT NULL,
+    dns_server_address_uuid TEXT NOT NULL,
+    CONSTRAINT fk_ip_address_dns_server_address_dns_server_address
+    FOREIGN KEY (dns_server_address_uuid)
+    REFERENCES ip_address (uuid),
+    CONSTRAINT fk_ip_address_dns_server_address_ip_address
+    FOREIGN KEY (address_uuid)
+    REFERENCES ip_address (uuid),
+    PRIMARY KEY (dns_server_address_uuid, address_uuid)
+);
+
+CREATE TABLE ip_address_dns_search_domain (
+    address_uuid TEXT NOT NULL,
+    search_domain TEXT NOT NULL,
+    CONSTRAINT fk_ip_address_dns_search_domain_ip_address
+    FOREIGN KEY (address_uuid)
+    REFERENCES ip_address (uuid),
+    PRIMARY KEY (address_uuid, search_domain)
+);
+
+CREATE TABLE fqdn_address (
+    uuid TEXT NOT NULL PRIMARY KEY,
+    hostname TEXT NOT NULL,
+    -- one of public, local-cloud, local-machine, link-local etc.
+    scope_id INT NOT NULL,
+
+    CONSTRAINT fk_fqdn_address_scope
+    FOREIGN KEY (scope_id)
+    REFERENCES network_address_scope (id)
+);
+
+CREATE UNIQUE INDEX idx_fqdn_address_hostname
+ON fqdn_address (hostname);
+
+CREATE TABLE net_node_fqdn_address (
+    net_node_uuid TEXT NOT NULL,
+    address_uuid TEXT NOT NULL,
+    CONSTRAINT fk_net_node_fqdn_address_net_node
+    FOREIGN KEY (net_node_uuid)
+    REFERENCES net_node (uuid),
+    CONSTRAINT fk_net_node_fqdn_address_address
+    FOREIGN KEY (address_uuid)
+    REFERENCES fqdn_address (uuid),
+    PRIMARY KEY (net_node_uuid, address_uuid)
+);
+
+-- v_address exposes ip and fqdn addresses as a single table.
+-- Used for compatibility with the current core network model.
+CREATE VIEW v_address AS
+SELECT
+    ipa.address_value,
+    ipa.type_id,
+    ipa.config_type_id,
+    ipa.origin_id,
+    ipa.scope_id
+FROM ip_address AS ipa
+UNION
+SELECT
+    fa.hostname AS address_value,
+    -- FQDN address type is always "hostname".
+    0 AS type_id,
+    -- FQDN address config type is always "manual".
+    3 AS config_type_id,
+    -- FQDN address origin is always "unknown".
+    0 AS origin_id,
+    fa.scope_id
+FROM fqdn_address AS fa;
+
+CREATE TABLE ip_address_provider (
+    -- a provider-specific ID of the IP address.
+    provider_id TEXT NOT NULL PRIMARY KEY,
+    address_uuid TEXT NOT NULL,
+    CONSTRAINT fk_provider_ip_address_ip_address
+    FOREIGN KEY (address_uuid)
+    REFERENCES ip_address (uuid)
+);
+
+CREATE TABLE ip_address_subnet (
+    subnet_uuid TEXT NOT NULL,
+    address_uuid TEXT NOT NULL,
+    CONSTRAINT fk_ip_address_subnet_subnet
+    FOREIGN KEY (subnet_uuid)
+    REFERENCES subnet (uuid),
+    CONSTRAINT fk_ip_address_subnet_ip_address
+    FOREIGN KEY (address_uuid)
+    REFERENCES ip_address (uuid),
+    PRIMARY KEY (subnet_uuid, address_uuid)
```

Comment:
This should be only the `address_uuid`. We're using a join table to avoid sparse data, but an address can only be in one subnet.